players/g4_player.py

```diff
@@ -158,30 +301,87 @@ def _reshape(
+        ameoba_cells = list(zip(*np.where(curr_state.amoeba_map == State.ameoba.value)))
         unoccupied_target_cells = target - set(ameoba_cells)
-
         to_occupy = set(occupiable_cells).intersection(unoccupied_target_cells)
 
-        return list(retractable_cells)[:len(to_occupy)], list(to_occupy), memory
+        k = min(len(to_occupy), len(retractable_cells))
+        retract = retract_k(k, list(retractable_cells), curr_state.amoeba_map)
+        extend = list(to_occupy)[:k]
+
+        # debug
+        visualize_reshape(
+            list(target), ameoba_cells,
+            occupiable_cells, list(retractable_cells),
+            retract, extend
+        )
+
+        return retract, extend, memory
+
+    #does COG need to be one of the ameoba cells?
+    def _get_cog(
+        self,
+        curr_state: AmoebaState,
+    ) -> tuple[int, int]:
+        """Compute center of gravity of current Ameoba"""
+        ameoba_cells = np.array(list(zip(*np.where(curr_state.amoeba_map == State.ameoba.value))))
+        cog = (round(np.average(ameoba_cells[:,0])),round(np.average(ameoba_cells[:,1])))
+        return cog
+    
+    def _get_xmax(self, curr_state: AmoebaState) -> int:
+        """Returns the x-value of the "rightmost" Ameoba cell.
+        
+        Note: when the ameoba's bucket arm moves from x=99 to the right, the x-value
+        will wrap around to 0. In this case, xmax is 0 since that's where the bucket
+        arms are. Below is a graphical illustration:
+
+                       x=0     ...     x = 98  x = 99
+                        .                 .      .
+                                          .      .
+                                          .      .
+                        .                 .      .
+                      (xmax)
+        """
+        ameoba_xs, _ = np.where(curr_state.amoeba_map == State.ameoba.value)
+        xmin, xmax = min(ameoba_xs), max(ameoba_xs)
+
+        # TODO: 5 is a magic number that should work... This is hacky though.
+        if xmax == 99 and xmin < 5:
+            xs = ameoba_xs[ameoba_xs <= 5]
+            return max(xs)
+
+        return xmax
+
+    def _in_shape(self, curr_state: AmoebaState) -> bool:
+        xmax = self._get_xmax(curr_state)
+        arms_expected = 1 + math.floor((curr_state.current_size - 3) / 7)
+        arms_got = len(np.where(curr_state.amoeba_map[xmax,:] == State.ameoba.value)[0])
+
+        return arms_got >= arms_expected
 
     def move(
-        self, state: AmoebaState, memory: int
+        self, prev_state: AmoebaState, state: AmoebaState, memory: int
     ) -> tuple[list[cell], list[cell], int]:
 
-        # TODO: compute ameoba's center of gravity, and xmax
-        size = 9
-        cog = (50, 50)
-        xmax = 51
+        size = (state.current_size)
+
+        # TODO: maybe not always shifting horizontally?
+        # cog = self._get_cog(state if np.any(prev_state.amoeba_map) else prev_state)
+        cog  = (50, 50)
```

Comment:
- y-value of `cog` : determines the **vertical center** of the ideal comb shape to morph the ameoba into
-  `xmax`: determines the x-value of the bucket arms

In hindsight, dynamically calculating $y_{cog}$ based on the current state causes issues: $y_{cog}$ could be different in consecutive turns, causing the positions of the bucket arm to shift up-and-down, as the following figure shows. So I'm hardcoding it to 50 ( $y_{cog}$ of the initial ameoba).

In the future, we could change $cog$ overtime to make the ameoba swift diagonally for example (something to explore in the future if needed).

![issue2](https://user-images.githubusercontent.com/25857014/203418725-8604afcb-6f08-440c-b5cf-a748cb1b495f.png)